### PR TITLE
Improve compatibility with Clang for Windows

### PR DESCRIPTION
**Pull request by Penguinwizzard**
This changeset removes some of the impediments to building ChakraCore with clang on windows. This is still not a supported configuration, just a less buggy one. There's still known failures with the staticinterpreterthunk in this configuration, and building the jit idl takes manual configuration of the relevant project file due to a clang bug.

### CODE REVIEW

File: lib/JITIDL/Chakra.JITIDL.vcxproj [hunk 1]
```diff
@@ -31,7 +31,9 @@
   <ItemGroup>
     <Midl Include="$(MsBuildThisFileDirectory)ChakraJIT.idl">
       <OutputDirectory>$(IntDir)</OutputDirectory>
-      <AdditionalOptions>%(AdditionalOptions) /acf "$(MsBuildThisFileDirectory)ChakraJIT.acf" /prefix client "Client" server "Server"</AdditionalOptions>
+      <!-- This is a messy work-around for a bug in clang and a bug in midl. Clang-cl purports to support a /P option to preprocess to a file, which does not work, so we have to find a different cl on the system. Midl makes us use short paths because it doesn't properly support spaces in arguments. -->
+      <AdditionalOptions Condition="'$(Clang)'!='ClangCL'">%(AdditionalOptions) /acf "$(MsBuildThisFileDirectory)ChakraJIT.acf" /prefix client "Client" server "Server"</AdditionalOptions>
+      <AdditionalOptions Condition="'$(Clang)'=='ClangCL'">%(AdditionalOptions) /acf "$(MsBuildThisFileDirectory)ChakraJIT.acf" /prefix client "Client" server "Server" /cpp_cmd "C:\Progra~2\MIB055~1\Preview\Enterprise\VC\Tools\MSVC\14.14.26310\bin\Hostx64\x64\cl.exe"</AdditionalOptions>
```
Comment (jackhorton) [inline lib/JITIDL/Chakra.JITIDL.vcxproj]: For my own understanding, where did this path come from?

File: test/runtests.py [hunk 2]
```diff
@@ -454,6 +454,9 @@ def _run_one_test(self, test):
                     args.flags.split() + \
                     flags.split()
 
+        if test.get('custom-config-file') != None:
+            flags = ['-CustomConfigFile:' + test.get('custom-config-file')]
```
Comment (jackhorton) [inline test/runtests.py]: This overwrites any flags set earlier in this function?

File: Build/Chakra.Build.Clang.props [hunk 3]
```diff
@@ -61,7 +61,9 @@
         -Wno-microsoft-extra-qualification
         -Wno-microsoft-default-arg-redefinition
         -Wno-microsoft-exception-spec
-        -v
+	-Wno-clang-cl-pch
```
Comment (jackhorton) [inline Build/Chakra.Build.Clang.props]: nit: spacing here?

File: lib/JITIDL/Chakra.JITIDL.vcxproj (hunk 1 above)
Comment (Penguinwizzard) [reply lib/JITIDL/Chakra.JITIDL.vcxproj]: This is very sloppy and the worst part of the PR. Currently this needs to be re-written on a per-system basis. Before any real attempt to merge this is made, I'm going to re-write this part, with the worst case being a script that attempts to figure out where the actual cl is. There's just too many bad constraints on this.

In terms of constructing the path, you can get windows short directory names with `dir /x`, which helps you work with tools that don't support spaces in arguments. It's a legacy artefact, but helps in situations like this one.

File: Build/Chakra.Build.Clang.props (hunk 3 above)
Comment (Penguinwizzard) [reply Build/Chakra.Build.Clang.props]: Yep, tab check caught this too, looks like my vim instance when run inside the vs dev tools console didn't pick up the default tab configuration...

**Comment by dilijev** [comment]
@Penguinwizzard are you still planning to merge this?

**Comment by Penguinwizzard** [comment]
It needs a little bit of clean-up with regards to how the midl handling is done, but yes. It's just not terribly high priority.

**Comment by Penguinwizzard** [comment]
Cleaned up and ready for review.

File: Build/Chakra.Build.Clang.Default.props [hunk 4]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="utf-8"?>
 <Project DefaultTargets="Build" ToolsVersion="12.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <!-- Not officially supported! Configuration to build with clang on windows -->
-  <PropertyGroup Condition="'$(Clang)'=='ClangCL'">
+  <PropertyGroup Condition="'$(Clang)'=='ClangCL' And '$(ForceMSVC)'==''">
     <PlatformToolset>LLVM-vs2014</PlatformToolset>
```
Comment (jackhorton) [inline Build/Chakra.Build.Clang.Default.props]: is vs2014 still correct?

Comment (Penguinwizzard) [reply Build/Chakra.Build.Clang.Default.props]: I believe that it is; the version packaged with VS matches this. There are GitHub repos with newer versions (llvm-vs2014 uses "clang with Microsoft CodeGen version 3.8.0", while there's GitHub repos for [5.0](https://github.com/arves100/llvm-vs2017-integration) and [6.0](https://github.com/arves100/llvm-vs2017-integration) third-party).

File: lib/JITIDL/Chakra.JITIDL.vcxproj (hunk 1 above)
Comment (Penguinwizzard) [reply lib/JITIDL/Chakra.JITIDL.vcxproj]: Found a much better work-around for the new version, thankfully.

**Pull request merged**